src/box/memtx_tx.c

```diff
@@ -3027,12 +3075,37 @@ memtx_tx_tuple_clarify_slow(struct txn *txn, struct space *space,
 } while (0)
 
 uint32_t
-memtx_tx_index_invisible_count_slow(struct txn *txn,
-				    struct space *space, struct index *index)
+memtx_tx_index_invisible_count_matching_until_slow(
+	struct txn *txn, struct space *space, struct index *index,
+	enum iterator_type type, const char *key, uint32_t part_count,
+	struct tuple *until, hint_t until_hint)
 {
+	struct key_def *key_def = index->def->key_def;
+	struct key_def *cmp_def = index->def->cmp_def;
+
+	/*
+	 * The border is only valid if it's located at or after the first
+	 * tuple in the index according to the iterator direction and key.
+	 */
+	assert(until == NULL ||
+	       memtx_tx_tuple_matches(key_def, until, type == ITER_EQ ?
+				      ITER_GE : type == ITER_REQ ? ITER_LE :
+				      type, key, part_count));
+
 	uint32_t res = 0;
 	struct memtx_story *story;
 	memtx_tx_foreach_in_index_tuple_story(space, index, story, {
+		/* Check the border using the cmp_def. */
+		if (until != NULL &&
+		    !memtx_tx_tuple_is_before(cmp_def, story->tuple,
+					      until, until_hint, type))
+			continue;
+
+		/* All tuples in the story chain share the same key. */
+		if (!memtx_tx_tuple_matches(key_def, story->tuple,
+					    type, key, part_count))
+			continue;
+
```

Comment:
Nit: It'd be nice if this was done in one helper function because this is the second place where you call them one after another.

Reply:
Done.